### PR TITLE
[JENKINS-60536] bugfix: Jenkins job hangs in Jira call after git rev-list command

### PR DESCRIPTION
**Pull request by bguerin**
All credits to https://github.com/xft-devs/jira-plugin

### CODE REVIEW

File: pom.xml [hunk 1]
```diff
@@ -21,6 +21,7 @@
     <java.level>8</java.level>
     <mockito.version>1.10.19</mockito.version>
     <jira-rest-client.version>5.2.1</jira-rest-client.version>
+    <powermock.version>1.6.3</powermock.version>
```
Comment (warden) [inline pom.xml:24]: This is not needed for this change

File: src/main/java/hudson/plugins/jira/JiraRestService.java [hunk 2]
```diff
@@ -294,7 +294,7 @@ public BasicIssue createIssue(String projectKey, String description, String assi
             // See upstream fix for setAssigneeName:
             //    https://bitbucket.org/atlassian/jira-rest-java-client/pull-requests/104/change-field-name-from-name-to-id-for/diff 
             builder.setFieldInput(new FieldInput(IssueFieldId.ASSIGNEE_FIELD, ComplexIssueInputFieldValue.with("accountId", assignee)));
-	    }
+        }
```
Comment (warden) [inline src/main/java/hudson/plugins/jira/JiraRestService.java:297]: This should not te touched.

**Comment by bguerin** [comment]
> Please rebase this to current master and squash the commits to contain only relevant changes).

done

File: src/main/java/com/atlassian/httpclient/apache/httpcomponents/ApacheAsyncHttpClient.java [hunk 3]
```diff
@@ -469,6 +477,18 @@ private void publishEvent(Request request, long requestDuration, Throwable ex)
 
     private PromiseHttpAsyncClient getPromiseHttpAsyncClient(Request request)
     {
+
+        log.info( "Creating new HttpAsyncClient" );
```
Comment (warden) [inline src/main/java/com/atlassian/httpclient/apache/httpcomponents/ApacheAsyncHttpClient.java:481]: Shouldn't this be `log.trace`?

File: src/main/java/com/atlassian/httpclient/apache/httpcomponents/ApacheAsyncHttpClient.java [hunk 4]
```diff
@@ -137,121 +137,128 @@ public ApacheAsyncHttpClient(final Supplier<String> applicationName,
...
                 }
+
+                clientBuilder.setRoutePlanner(
+                    new JenkinsProxyRoutePlanner( proxy, proxyConfiguration.getNoProxyHostPatterns() ) );
             }
+        }
 
-            /*
-            ProxyConfigFactory.getProxyHost(options).foreach(new Effect<HttpHost>()
+        /*
```
Comment (warden) [inline src/main/java/com/atlassian/httpclient/apache/httpcomponents/ApacheAsyncHttpClient.java:242]: Probably this commented out code could be removed.

**Comment by bguerin** [comment]
> only thing is that it could use some test.. but I'm not sure how feasible it is to create one easily for this scenario.

tests added, for what they are worth

File: src/test/java/com/atlassian/httpclient/apache/httpcomponents/SettableFuturePromiseHttpPromiseAsyncClientTest.java [hunk 5]
```diff
@@ -0,0 +1,107 @@
...
+    private HttpResponse response;
+
+    @Mock
+    private HttpContext context;
+
+    @InjectMocks
+    private SettableFuturePromiseHttpPromiseAsyncClient<Object> asyncClient;
+
+    @Test
+    public void ensure_close_httpclient_on_completion() throws IOException
```
Comment (artkoshelev) [inline src/test/java/com/atlassian/httpclient/apache/httpcomponents/SettableFuturePromiseHttpPromiseAsyncClientTest.java:55]: pls use camelCase everywhere

Comment (bguerin) [reply src/test/java/com/atlassian/httpclient/apache/httpcomponents/SettableFuturePromiseHttpPromiseAsyncClientTest.java:55]: ok, sorry, I took ApacheAsyncHttpClientTest as example

**Comment by warden** [comment]
@bguerin great job, thanks for that!

**Comment by bguerin** [comment]
welcome :)

**Comment by warden** [comment]
@olamy [there's a need for releasing this apparently](https://issues.jenkins-ci.org/browse/JENKINS-60536). Do you have some spare time or want me to do it? 
Because of #239 I vote this to be 3.1.0.

**Comment by olamy** [comment]
@warden fine I will do it. version number not sure as there is no visible changes or API change for users. But I don't really care 😄

**Pull request merged**